packages/merkleize-metadata/src/main.ts

```diff
@@ -0,0 +1,191 @@
+import {
+  Blake3256,
+  HexString,
+  compact,
+  u32,
+} from "@polkadot-api/substrate-bindings"
+import {
+  ExtraInfo,
+  ExtrinsicMetadata,
+  LookupValue,
+  MetadataDigest,
+  ScalePrimitive,
+  TypeRef,
+  extraInfo,
+  extrinsicMetadata,
+  lookupType,
+  metadataDigest,
+  extrinsicDec,
+} from "./codecs"
+import { getAccessibleTypes } from "./get-accessible-types"
+import { getLookup } from "./get-lookup"
+import { getMetadata } from "./get-metadata"
+import { compactTypeRefs, mergeUint8, toBytes } from "./utils"
+import { decodeAndCollectKnownLeafs } from "./decode-and-collect"
+import { getProofData } from "./proof"
+
+export const merkleizeMetadata = (
+  metadataBytes: Uint8Array | HexString,
+  info: ExtraInfo,
+) => {
+  const metadata = getMetadata(metadataBytes)
+
+  const definitions = new Map<number, LookupValue>(
+    metadata.lookup.map((value) => [value.id, value]),
+  )
+  const accessibleTypes = getAccessibleTypes(metadata, definitions)
+  const getPrimitive = (frameId: number): ScalePrimitive["tag"] | null => {
+    const {
+      def: { tag, value },
+    } = definitions.get(frameId)!
+
+    if (tag === "primitive") return value.tag
+
+    if ((tag !== "composite" && tag !== "tuple") || value.length > 1)
+      throw new Error("The provided definition doesn't map to a primitive")
+
+    return value.length === 0
+      ? null // signals `void`
+      : getPrimitive(tag === "tuple" ? value[0] : value[0].type)
+  }
+
+  const getTypeRef = (frameId: number): TypeRef => {
+    const { def } = definitions.get(frameId)!
+
+    if (def.tag === "primitive") return { tag: def.value.tag, value: undefined }
+
+    if (def.tag === "compact") {
+      const primitive = getPrimitive(def.value)
+      const tag = compactTypeRefs[primitive as "u8"]
+      if (!tag) throw new Error("Invalid primitive for Compact")
+      return { tag, value: undefined }
+    }
+
+    return accessibleTypes.has(frameId)
+      ? { tag: "perId", value: accessibleTypes.get(frameId)! }
+      : { tag: "void", value: undefined }
+  }
+
+  const extrinsic: ExtrinsicMetadata = {
+    version: metadata.extrinsic.version,
+    addressTy: getTypeRef(metadata.extrinsic.address),
+    callTy: getTypeRef(metadata.extrinsic.call),
+    signatureTy: getTypeRef(metadata.extrinsic.signature),
+    signedExtensions: metadata.extrinsic.signedExtensions.map((se) => ({
+      identifier: se.identifier,
+      includedInExtrinsic: getTypeRef(se.type),
+      includedInSignedData: getTypeRef(se.additionalSigned),
+    })),
+  }
+
+  const lookup = getLookup(
+    definitions,
+    accessibleTypes,
+    getTypeRef,
+    getPrimitive,
+  )
+  const lookupEncoded = lookup.map(lookupType.enc)
+
+  let _proofs: Array<Uint8Array> | undefined
+  const getProofs = (): Array<Uint8Array> => {
```

Comment:
I'd rather rename this to `getHashTree` or something similar. They are not really proofs, just the hash tree.
I confused this with `getProofData` when reviewing the code.